### PR TITLE
Implement host and view injectors

### PR DESCRIPTION
**Pull request by vsavkin**

### CODE REVIEW

File: modules/angular2/test/core/compiler/element_injector_spec.js [hunk 1]
```diff
@@ -285,13 +272,13 @@ export function main() {
...
-          - c2
-        - p2
-          - c3
+     Build a tree of the following shape:
+     root
+     - p1
+     - c1
+     - c2
+     - p2
+     - c3
```
Comment (vicb) [inline modules/angular2/test/core/compiler/element_injector_spec.js]: oops formatting

**Comment by vicb** [comment]
Could you add a `BREAKING CHANGES` section to the commit message for the changelog ?

File: modules/angular2/test/core/compiler/integration_spec.js [hunk 2]
```diff
@@ -823,6 +823,42 @@ export function main() {
       }));
     });
 
+    describe("dependency injection", () => {
```
Comment (tbosch) [inline modules/angular2/test/core/compiler/integration_spec.js]: Add integration test for @Unbounded

File: modules/angular2/src/core/annotations_impl/annotations.js [hunk 3]
```diff
@@ -660,6 +660,9 @@ export class Directive extends Injectable {
   //TODO(vsavkin): This would better fall under the Macro directive concept.
   compileChildren: boolean;
 
+  viewInjector:List;
```
Comment (tbosch) [inline modules/angular2/src/core/annotations_impl/annotations.js]: Add comments with little examples (see the other properties)

File: modules/angular2/src/core/annotations_impl/visibility.js [hunk 4]
```diff
@@ -103,9 +126,16 @@ export class Parent extends DependencyAnnotation {
...
-export class Ancestor extends DependencyAnnotation {
+export class Ancestor extends Visibility {
   @CONST()
   constructor() {
-    super();
+    super(999999, false);
+  }
+}
+
+export class Unbounded extends Visibility {
```
Comment (tbosch) [inline modules/angular2/src/core/annotations_impl/visibility.js]: Add `exportAs` doc

File: modules/angular2/src/core/compiler/element_injector.js [hunk 5]
```diff
@@ -225,20 +222,19 @@ export class DirectiveDependency extends Dependency {
 }
 
 export class DirectiveBinding extends ResolvedBinding {
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/element_injector.js]: Maybe add a `TODO: move this out of the `ElementInjector``

File: modules/angular2/src/core/compiler/element_injector.js [hunk 6]
```diff
@@ -431,14 +417,31 @@ export class ProtoElementInjector  {
   /** The variable name that will be set to $implicit for the element. */
   exportImplicitName:string;
 
-  constructor(parent:ProtoElementInjector, index:int, bindings:List, firstBindingIsComponent:boolean = false, distanceToParent:number = 0) {
+  _shadowDomRoot:ProtoElementInjector;
+
+  static create(parent:ProtoElementInjector, index:int, bindings:List, hasShadowRoot:boolean, distanceToParent:number) {
+    var hostBindings = [];
+    ListWrapper.forEach(bindings, b => {
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/element_injector.js]: Extract into help function

File: modules/angular2/src/core/compiler/element_injector.js [hunk 7]
```diff
@@ -431,14 +417,31 @@ export class ProtoElementInjector  {
...
+      ListWrapper.push(hostBindings, b);
+      ListWrapper.forEach(b.resolvedHostInjectables, binding => ListWrapper.push(hostBindings, binding));
+    });
+
+    var pei = new ProtoElementInjector(parent, index, hostBindings, distanceToParent);
+    if (hasShadowRoot) {
+      var shadowRootEi = new ProtoElementInjector(null, index, bindings[0].resolvedViewInjectables, 0);
+      pei._shadowDomRoot = shadowRootEi;
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/element_injector.js]: Add TODO: maybe create special class for `shadowDomRoot` injector as we only use the bindings but not other things like the `host` relationship or the app injector.

File: modules/angular2/src/core/compiler/element_injector.js [hunk 8]
```diff
@@ -391,17 +380,16 @@ ElementInjector:
...
-  _binding2:DirectiveBinding;
-  _binding3:DirectiveBinding;
-  _binding4:DirectiveBinding;
-  _binding5:DirectiveBinding;
-  _binding6:DirectiveBinding;
-  _binding7:DirectiveBinding;
-  _binding8:DirectiveBinding;
-  _binding9:DirectiveBinding;
-  _binding0IsComponent:boolean;
+  _binding0:ResolvedBinding;
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/element_injector.js]: Add a comment: This is the component binding if this is a component.

File: modules/angular2/src/core/compiler/element_injector.js [hunk 9]
```diff
@@ -1011,6 +995,13 @@ export class ElementInjector extends TreeNode {
     }
   }
 
+  _traverseUp(ei, visibility) {
+    if (isPresent(ei._parent)) return ei._parent;
+    if (isPresent(ei._host) && isPresent(ei._host._shadowDomRoot) && ei._host._shadowDomRoot !== ei) return ei._host._shadowDomRoot;
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/element_injector.js]: Try to use `parent` relationship between view root eli and shadowDomRoot.

File: modules/angular2/src/di/injector.ts [hunk 10]
```diff
@@ -96,6 +96,10 @@ export class Injector {
     return _createListOfBindings(flatten);
   }
 
+  static resolve2(bindings: List<any>): List<ResolvedBinding> {
```
Comment (tbosch) [inline modules/angular2/src/di/injector.ts]: Merge with normal `resolve`

File: modules/angular2/test/core/compiler/element_injector_spec.js [hunk 11]
```diff
@@ -629,81 +581,110 @@ export function main() {
...
+          expect(destroy.onDestroyCounter).toBe(1);
+        });
 
-      it("should publish component to its children via app injector when requested", function() {
-        var parentDirective = new Component({
-          selector: 'parent',
-          publishAs: [ParentInterface]
+        it("should work with services", function() {
+          var inj = injector([DirectiveBinding.createFromType(SimpleDirective, new DummyDirective({hostInjector: [SimpleService]}))]);
+          inj.dehydrate();
```
Comment (tbosch) [inline modules/angular2/test/core/compiler/element_injector_spec.js]: Maybe get the service out of the injector and check the type

File: modules/angular2/src/core/compiler/element_injector.js [hunk 12]
```diff
@@ -266,59 +262,52 @@ export class DirectiveBinding extends ResolvedBinding {
   }
 
   get changeDetection() {
-    if (isPresent(metadata)) {
-      return metadata.changeDetection;
-    } else {
-      return null;
-    }
+    return isPresent(metadata) ? metadata.changeDetection : null;
```
Comment (TommyM) [inline modules/angular2/src/core/compiler/element_injector.js]: Should this not be this.metadata?

**Pull request merged**